### PR TITLE
feat: gnovm benchmarking tool

### PR DESCRIPTION
**Pull request by piux2**
<details><summary>Contributors' checklist...</summary>

- [ x] Added new tests, or not needed, or not feasible
- [ x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ x] Updated the official documentation or not needed
- [ x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ x] Added references to related issues and PRs
- [ x] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

We build this tool mainly for the following issues

https://github.com/gnolang/gno/issues/1826
https://github.com/gnolang/gno/issues/1828
https://github.com/gnolang/gno/issues/1281
https://github.com/gnolang/gno/issues/1973

We could also use it in the following cases. 

https://github.com/gnolang/gno/issues/1973
https://github.com/gnolang/gno/issues/2222

### Minimum Overhead and Footprint

- Constant build flags enable benchmarking.
- Encode operations and measurements in binary.
- Dump to a local file in binary.
- No logging, printout, or network access involved.

### Accuracy

- Pause the timer for storage access while performing VM opcode benchmarking.
- Measure each OpCode execution in nanoseconds.
- Store access includes the duration for Amino marshalling and unmarshalling.

It is built on top of @deelawn's design and framework with @jaekwon's input. 
https://github.com/gnolang/gno/pull/2073

### Simple mode

The benchmark only involves the GnoVM and the persistent store. It benchmarks the bare minimum components, and the results are isolated from other components. We use standardize gno contract to perform the benchmarking.

This mode is the best for benchmarking each major release and/or changes in GnoVM.

    make opcode
    make storage

### Production mode

It benchmarks the node in the production environment with minimum overhead.
We can only benchmark with standardize the contract but also capture the live usage in production environment.
It gives us a complete picture of the node perform.

  1. Build the production node with benchmarking flags:

  `go build -tags "benchmarkingstorage benchmarkingops" gno.land/cmd/gnoland`

  2. Run the node in the production environment. It will dump benchmark data to a benchmark.bin file.

  3. call the realm contracts at `gno.land/r/x/benchmark/opcodes` and `gno.land/r/x/benchmark/storage`

  4. Stop the server after the benchmarking session is complete.

  5. Run the following command to convert the binary dump:

  `gnobench -bin path_to_benchmark_bin`

    it converts the binary dump to results.csv and results_stats.csv.

## Results ( Examples )

The benchmarking results are stored in two files:
  1. The raw results are saved in results.csv.

  | Operation       | Elapsed Time | Disk IO Bytes |
  |-----------------|--------------|---------------|
  | OpEval          | 40333        | 0             |
  | OpPopBlock      | 208          | 0             |
  | OpHalt          | 167          | 0             |
  | OpEval          | 500          | 0             |
  | OpInterfaceType | 458          | 0             |
  | OpPopBlock      | 166          | 0             |
  | OpHalt          | 125          | 0             |
  | OpInterfaceType | 21125        | 0             |
  | OpEval          | 541          | 0             |
  | OpEval          | 209          | 0             |
  | OpInterfaceType | 334          | 0             |

  2. The averages and standard deviations are summarized in results_stats.csv.

  | Operation      | Avg Time | Avg Size | Time Std Dev | Count |
|----------------|----------|----------|--------------|-------|
| OpAdd          | 101      | 0        | 45           | 300   |
| OpAddAssign    | 309      | 0        | 1620         | 100   |
| OpArrayLit     | 242      | 0        | 170          | 700   |
| OpArrayType    | 144      | 0        | 100          | 714   |
| OpAssign       | 136      | 0        | 95           | 2900  |
| OpBand         | 92       | 0        | 30           | 100   |
| OpBandAssign   | 127      | 0        | 62           | 100   |
| OpBandn        | 97       | 0        | 54           | 100   |
| OpBandnAssign  | 125      | 0        | 113          | 100   |
| OpBinary1      | 128      | 0        | 767          | 502   |
| OpBody         | 127      | 0        | 145          | 13700 |

### CODE REVIEW

**Review by deelawn — APPROVED** [review]
Looks good.  Perhaps we could move the transient `size` variable so that is defined inside the benchmarking package as an exported variable.  Then it can be set and read strictly within benchmarking constant conditional blocks and avoid any unnecessary allocations when benchmarking is not enabled.

What do you think?

File: benchmarking/Makefile
Comment (thehowl) [inline benchmarking/Makefile:1]: Could this be reorganised to:

- `gnovm/cmd/gnobench` + `gnovm/pkg/gasbench` OR
- `contribs/gnobecnh`?

I don't want to add another top-level directory.

**Comment by jaekwon** [comment]
Waiting for updates as per discussions with Ray

File: benchmarking/Makefile
Comment (jaekwon) [reply benchmarking/Makefile:1]: gnovm/pkg/opsbench

**Review by ltzmaxwell — COMMENTED** [review]
preliminary review.

File: gnovm/pkg/benchops/bench.go
Comment (ltzmaxwell) [inline gnovm/pkg/benchops/bench.go:60]: unset `isOpCodeStarted`?

**Comment by zivkovicmilos** [comment]
Hey @piux2, can you rebase this with `master`? 🙏

**Review by mvertes — COMMENTED** [review]
The pr provides:

1. Performance measurement capabilities for opcodes (count, execution time) and
persistent storage (count, size). This part is similar to performance counter
registers in processors. It would better be called `perf` (the VM capability)
instead of `benchops` (a particular usage of this capability, and not the only possible one, i.e    profiling, gas measurement).
See for example [linux](https://perfwiki.github.io/main/) or [freebsd](https://man.freebsd.org/cgi/man.cgi?query=pmc&sektion=3&apropos=0&manpath=FreeBSD+14.1-RELEASE+and+Ports)

2. An executable called `gnobench`, to perform a benchmark on a specific scenario (a sample forum   board in gno). This executable should not be located under gnovm, but in `contribs/gnobench`, with
its dependencies:
  - the readme, makefile, etc in `contrib/benchmark`
  - the executable itself in `contrib/benchmark/cmd/gnobench`
  - benchmarked gno files in `contrib/benchmark/gno`
  - the vm part still in `gnovm/pkg/perf`

Regarding the build tags to enable or not the counting, I'm not a big fan of that. It brings a lot  of hassle, and it's partly useless because counting is necessary in production anyway. The          performance gain vs using a variable instead of a constant (requiring a different executable) to enable the feature in opcode procressing was shown negligible when we introduced the debugger, it should be the same here. A former colleague of mine used to say: "Premature optimisation is the root of all evil" :-)

In more details, the measurement of opcodes consists of counting and measuring execution time. Counting is super cheap (a counter increment operation), and could be always performed. Measuring   execution time is probably a bit more costly and it makes sense to enable it only for benchmark scenarios. In production, we could have a static pre-computed profile of mean exec time per opcode, for gas estimation. But again, no sufficient reason to use build tags.

Regarding the storage of performance counters, they should be stored per virtual machine, instead of a global variable `measure`, which is generally a bad practice, bringing a lot of unstated constraints and risks. I understand that it is already the case for other parts of the VM, and that it facilitates the coding, but this is something that we should eventually fix. It can be done later, I don't want to block the PR for that.

File: gnovm/pkg/gnolang/store.go
Comment (piux2) [reply gnovm/pkg/gnolang/store.go:580]: This was purposely left out. GetType(id) is part of a recursive call flow through the realm's fillType() function. To provide an accurate benchmarking measurement, we moved away from the stack-based measurement approach, which involved pushing and pausing a timer on the stack when the runtime entered a function call. The trade-off is that we can no longer measure the runtime of functions involved in recursive calls. For now, we can assume and assign the same gas value to GetType as we do for SetType.

File: gnovm/pkg/benchops/bench.go
Comment (piux2) [reply gnovm/pkg/benchops/bench.go:53]: good catch!

File: gnovm/pkg/benchops/bench.go
Comment (piux2) [reply gnovm/pkg/benchops/bench.go:60]: Good catch! I will fix it.

**Comment by Gno2D2** [comment]
I'm a bot that assists the Gno Core team in maintaining this repository. My role is to ensure that contributors understand and follow our guidelines, helping to streamline the development process.

The following requirements must be fulfilled before a pull request can be merged.
Some requirement checks are automated and can be verified by the CI, while others need manual verification by a staff member.

These requirements are defined in this [configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

## Automated Checks

 🟢 Maintainers must be able to edit this pull request
 🔴 The pull request head branch must be up-to-date with its base


## Manual Checks

*No manual checks match this pull request.*

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Head branch (piux2:feat_gnobench) is up to date with base (master): behind by 128 / ahead by 18

```
</blockquote></details>

</blockquote></details>



</blockquote></details>

**Comment by piux2** [comment]
> Hey @piux2, can you rebase this with `master`? 🙏

done

**Comment by piux2** [comment]
@mvertes  thank you for the comments!

> The pr provides:
> 
> 1. Performance measurement capabilities for opcodes (count, execution time) and persistent storage (count, size). This part is similar to performance counter registers in processors. It would better be called `perf` (the VM capability) instead of `benchops` (a particular usage of this capability, and not the only possible one, i.e profiling, gas measurement). See for example [linux](https://perfwiki.github.io/main/) or [freebsd](https://man.freebsd.org/cgi/man.cgi?query=pmc&sektion=3&apropos=0&manpath=FreeBSD+14.1-RELEASE+and+Ports)

Thank you for the reference.!

I can see that Linux's perf tool measures software and hardware events in a way similar to how we measure opcode execution and storage access times. However, in our case, we use a set of standardized contracts in benchops/gno to evaluate the execution time of each opcode in the Gno VM runtime. This forms the essence of our benchmarking.

To truly implement performance measurement of the VM, similar to what Linux perf does with software and hardware events, we would need to develop a comprehensive new tool.

Alternatively, we might reserve the name 'Gno Performance and Profiling Tools' for measuring the VM's execution of user contracts in terms of gas consumption, similar to how the Go runtime measures performance for Go code.

> 2. An executable called `gnobench`, to perform a benchmark on a specific scenario (a sample forum   board in gno). This executable should not be located under gnovm, but in `contribs/benchmark`, with
>    its dependencies:
>    
>    * the readme, makefile, etc in `contrib/benchmark`
>    * the executable itself in `contrib/benchmark/cmd/gnobench`
>    * benchmarked gno files in `contrib/benchmark/gno`
>    * the vm part still in `gnovm/pkg/perf`
> 

I moved benchops/cmd out of /pkg and placed it in gnovm/cmd/benchops since it is still part of the VM.

> Regarding the build tags to enable or not the counting, I'm not a big fan of that. It brings a lot of hassle, and it's partly useless because counting is necessary in production anyway. The performance gain vs using a variable instead of a constant (requiring a different executable) to enable the feature in opcode procressing was shown negligible when we introduced the debugger, it should be the same here. A former colleague of mine used to say: "Premature optimisation is the root of all evil" :-)
> 

I totally agree  and advocate "Premature optimization is the root of all evil".  

All we did was using build flag to set constant flag instead of a variable flag to determine non-production code. Go complier removes entire if block if the condition is a constant value. The constant flag removes the over head 10-18% when we try to capture the accurate value of the operation in nano second with a timer

**Review by thehowl — APPROVED** [review]
Made a few changes for consistency and usability of the tool, but I'd say we're good to merge and iterate.

**Pull request merged**
